### PR TITLE
Storage: Improve pool name and volume name validation

### PR DESCRIPTION
**Pull request by tomponline**

### CODE REVIEW

File: lxd/storage/drivers/utils.go [hunk 1]
```diff
@@ -895,3 +896,45 @@ func remoteVolumeMapLock(connectorName string, driverName string) (locking.Unloc
...
+
+// ValidVolumeName validates a volume name.
+func ValidVolumeName(volumeName string) error {
+	if volumeName == "" {
+		return errors.New("Cannot be empty")
+	}
+
+	if strings.Contains(volumeName, "\\") {
+		return errors.New("Cannot contain backslashes")
+	}
```
Comment (simondeziel) [inline lxd/storage/drivers/utils.go:939]: Don't we need the same in `ValidPoolName()`?

Comment (simondeziel) [reply lxd/storage/drivers/utils.go:939]: Also for both, should we not block `..` like some open coded validation did before?

File: lxd/daemon_storage.go [hunk 2]
```diff
@@ -146,14 +146,14 @@ func daemonStorageSplitVolume(volume string) (poolName string, volumeName string
...
-	// Validate pool name.
-	if strings.Contains(poolName, "\\") || strings.Contains(poolName, "..") {
-		return "", "", fmt.Errorf("Invalid pool name: %q", poolName)
+	err = storageDrivers.ValidPoolName(poolName)
+	if err != nil {
+		return "", "", fmt.Errorf("Invalid pool name %q: %w", poolName, err)
 	}
 
-	// Validate volume name.
-	if strings.Contains(volumeName, "\\") || strings.Contains(volumeName, "..") {
```
Comment (simondeziel) [inline lxd/daemon_storage.go:155]: The `..` should still be banned, isn't it?

File: lxd/storage/drivers/utils.go [hunk 3]
```diff
@@ -903,7 +903,11 @@ func ValidPoolName(poolName string) error {
 		return errors.New("Cannot be empty")
 	}
 
-	if strings.Contains(poolName, shared.SnapshotDelimiter) {
+	if strings.HasPrefix(poolName, "-") {
```
Comment (simondeziel) [inline lxd/storage/drivers/utils.go:910]: Should we do the same for `ValidVolumeName()`?

File: lxd/storage/drivers/utils.go [hunk 4]
```diff
@@ -903,7 +903,11 @@ func ValidPoolName(poolName string) error {
...
+		return errors.New("Cannot start with a hyphen")
+	}
+
+	if strings.Contains(poolName, "/") {
```
Comment (simondeziel) [inline lxd/storage/drivers/utils.go]: Should we do the same for `ValidVolumeName()`?

File: test/suites/storage.sh [hunk 5]
```diff
@@ -7,10 +7,17 @@ test_storage() {
...
-  # edit storage and pool description
   local storage_pool storage_volume
   storage_pool="lxdtest-$(basename "${LXD_DIR}")-pool"
   storage_volume="${storage_pool}-vol"
+
+  # Check for pool name validation
+  ! lxc query -X POST --wait /1.0/storage-pools --data '{\"driver\":\"'"${lxd_backend}"'\"}' || false
+  ! lxc query -X POST --wait /1.0/storage-pools --data '{\"name\":\"-'"${storage_pool}"'\", \"driver\":\"'"${lxd_backend}"'\"}' || false
+  ! lxc storage create "${storage_pool}/" "$lxd_backend" || false
+  ! lxc storage create "${storage_pool} " "$lxd_backend" || false
```
Comment (simondeziel) [inline test/suites/storage.sh]: Nit pick: this new occurrence could use `{}`: `${lxd_backend}`
I know we are absolutely not consistent about this rule anywhere but since you've done it for the other variables :)

File: lxd/storage/drivers/utils.go (hunk 1 above)
Comment (tomponline) [reply lxd/storage/drivers/utils.go:939]: > Don't we need the same in `ValidPoolName()`?

Thats outside of the scope of this change and may introduce further potential breakages.

File: lxd/daemon_storage.go (hunk 2 above)
Comment (tomponline) [reply lxd/daemon_storage.go:155]: The volume name policy should be applied consistently throughout he code base.
But only here was ".." blocked.
I'm not sure if blocking `/` is sufficient for `..` to not be a concern.

File: lxd/storage/drivers/utils.go (hunk 3 above)
Comment (tomponline) [reply lxd/storage/drivers/utils.go:910]: Its outside of the scope of this particular PR as I'm aiming to fix the specific issue without introducing the risk of too many potential regressions.

File: lxd/storage/drivers/utils.go (hunk 4 above)
Comment (tomponline) [reply lxd/storage/drivers/utils.go]: yes we do with `if strings.Contains(volumeName, shared.SnapshotDelimiter) {`

File: lxd/storage/drivers/utils.go (hunk 3 above)
Comment (simondeziel) [reply lxd/storage/drivers/utils.go:910]: OK but in the long run blocking `-` would be a safe thing to do considering how easy to turn the name into a command argument.

File: lxd/storage/drivers/utils.go (hunk 4 above)
Comment (tomponline) [reply lxd/storage/drivers/utils.go]: oh right i already fixed this once in an earlier commit and then git reintroduced it again later

File: test/suites/storage.sh (hunk 5 above)
Comment (tomponline) [reply test/suites/storage.sh]: I hate bash so much, but will fix.

File: lxd/daemon_storage.go (hunk 2 above)
Comment (simondeziel) [reply lxd/daemon_storage.go:155]: What about a pool named *just* `..`?

File: lxd/storage/drivers/utils.go (hunk 3 above)
Comment (tomponline) [reply lxd/storage/drivers/utils.go:910]: It needs a full impact audit though - i wont include it in this PR, but if I get chance i'll evaluate it.

Whilst im fully onboard with the change, we do have to consider regressions with any policy change.

File: lxd/daemon_storage.go (hunk 2 above)
Comment (tomponline) [reply lxd/daemon_storage.go:155]: yeah ill add a check for that and volume name as should be very low impact.

**Comment by simondeziel** [comment]
> **lxd/storage/drivers/utils.go:900**
> 
>     * Consider adding unit tests for `ValidPoolName` and `ValidVolumeName` to verify all edge cases (empty, whitespace, leading hyphens, "..", backslashes, etc.).

I do agree with CodeQL here but that's material for another PR probably.

File: lxd/storage/drivers/utils.go (hunk 3 above)
Comment (simondeziel) [reply lxd/storage/drivers/utils.go:910]: `;` would also be another candidate to block. Maybe `{`, `}`, `|` and `&`...

It's a bit late to try to retrofit those restrictions :/

**Comment by tomponline** [comment]
> > **lxd/storage/drivers/utils.go:900**
> > ```
> > * Consider adding unit tests for `ValidPoolName` and `ValidVolumeName` to verify all edge cases (empty, whitespace, leading hyphens, "..", backslashes, etc.).
> > ```
> 
> I do agree with CodeQL here but that's material for another PR probably.

yep wouldn't hurt, at least we've got integration tests that cover the pool name stuff.

File: lxd/storage/drivers/utils.go (hunk 3 above)
Comment (tomponline) [reply lxd/storage/drivers/utils.go:910]: indeed - baby steps

Comment (tomponline) [reply lxd/storage/drivers/utils.go:910]: @simondeziel tested creating volumes starting with `-` on dir, zfs, ceph, lvm and btrfs, and all work fine because LXD prefixes the project name to the volume name.
